roles/ipsec/templates/cluster.conf.j2

```diff
@@ -0,0 +1,14 @@
+{% for host in groups['all'] %}
+{% if host != inventory_hostname %}
+conn {{ host }}
+        authby=rsasig
+        left={{ ip  }}
+        leftid=@{{ansible_hostname }}
+        right={{ hostvars[host]['ip'] }}
+        rightid=@{{ hostvars[host]['ansible_hostname'] }}
+{{ ipsec_left.stdout }}
+{{ hostvars[host]['ipsec_right']['stdout'] }}
```

Comment:
This value can't be cached and isn't delegated. It will break the ability to use `--limit newhosts` because this will require all hosts to be in the current play.

The only way to enable play limits with this reference is to add a task that does something like this:
```
- name: register ipsec_right
  shell: |
     docker run -v /etc/ipsec.d:/etc/ipsec.d cornelius/docker-ipsec-libreswan   \
              ipsec showhostkey --right --ckaid {{ ipsec_ckaid.stdout }}
  register: ipsec_right
  become: yes
  delegate_to: "{{ item }}"
  with_items: "groups['k8s-cluster']|union(groups['etcd']|union(groups['calico-rr']|default([])|unique|sort"
```

It's certainly not ideal, but it should work

Reply:
I have to think about this, because the connection should be updated always on both sides.
So I don't know if delegate will work here.

Reply:
What does the ID look like? Can the ID be predictable and based on something like hostname or ip address?

Reply:
No it is like an individual host key, like the ones used with ssh.
For the pure templating your suggestion is working, but might result in unencrypted or not working at all connections between two nodes.

I will think about this and try to find a solution.

Reply:
Another two options are:
use `fetch` module and aggregate them in a dir with the path credentials/ipsec/right-{{ inventory_hostname }}. Then you can use a file lookup to insert it into any command or loop through it here.
The other choice is generate the certs in advance on the ansible host with a custom command or the password lookup module

Reply:
The problem is not that I cannot get this via delegate, it will work.
The problem is that the ipsec connection is bidirectional and it needs to be configiured on both sides. So if set of hosts in the play was limited it leads to configurations only changed at one side of the connection. This problem I need to think about.

Reply:
Hi @mattymo, while working on this I quickly noticed that the currently the playbooks don't match the requirement to be able to run them on limited hosts. If I try to do that it fails for me quite fast in the kubernetes/preinstall role:
```fatal: [k8s-03]: FAILED! => {
    "failed": true, 
    "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'dict object' has no attribute 'ansible_default_ipv4'\n\nThe error appears to have been in '/home/jck/git/kubespray/roles/kubernetes/preinstall/tasks/set_facts.yml': line 14, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- set_fact:\n  ^ here\n"
}
```
I think that [setup with delegated facts](http://docs.ansible.com/ansible/latest/playbooks_delegation.html#delegated-facts) could be a solution for that Problem and also help me in my case. However it seems to be [broken in current anslible versim](https://github.com/ansible/ansible/issues/20980) and I cannot get it running.

So I am still working on it but for now I don't find an elegant solution.
Just to keep you up to date with what I was doing.

Reply:
@cornelius-keller Must this ipsec implementaion work in a full mesh mode? Is there no hub and spoke approach? For example, kube-master nodes have a bidirectional link to all kube-nodes, but no two kube-nodes have a bidirectional tunnel. All their node to node traffic would pass through a kube-master hode instead. 

If it can't be done, then we will go forward with your initial proposal and document the requirement that every node needs to set up a link with every other node.